Cache manifest everywhere

Hexcles opened the pull request:
Move the cache layer of manifest objects from wpt.testfiles to
manifest.manifest so that all users of the manifest module can benefit.

This prevents us from having two in-memory copies of the manifest when
running affected tests (one loaded by wpt.testfiles and the other loaded
by wptrunner.testloader). Partly fixes #14421 (the memory usage part of
it, but not the code health part).

Hexcles committed:
Cache manifest everywhere

Move the cache layer of manifest objects from wpt.testfiles to
manifest.manifest so that all users of the manifest module can benefit.

This prevents us from having two in-memory copies of the manifest when
running affected tests (one loaded by wpt.testfiles and the other loaded
by wptrunner.testloader). Partly fixes #14421 (the memory usage part of
it, but not the code health part).

Hexcles commented on tools/wpt/testfiles.py:
```diff
@@ -183,24 +181,12 @@ def _in_repo_root(full_path):
-
-load_manifest = _init_manifest_cache()
+def load_manifest(manifest_path=None, manifest_update=True):
+    from manifest import manifest
```
Switch to a delayed import here because otherwise Python will think there are two different `manifest` modules (because of different import paths) and we don't get the singleton.

jugglinmike replied:
What do you say we document your rationale with a comment? I'm afraid of your
optimization being accidentally subverted by some future refactoring (the
corresponding performance degradation might not be noticed immediately).

This is beyond my experience with Python, though. Can you advise on any
longer-term goals that would obviate this kind of consideration?

Hexcles replied:
I added a comment here, but I don't really have great ideas how to make sure we always get a singleton.

jgraham replied:
Did you verify that the previous code was broken? I know that `foo.bar` and `bar` are different even if they're the same `bar.py` file, but I'm slightly surprised if that also affects relative imports.

Hexcles commented:
On my machine, this reduces 2GB of memory usage (peak memory 5GB -> 3GB) when running

```bash
./wpt run --affected HEAD^1 --log-mach - --binary ~/tools/firefox-nightly/firefox firefox
```
(with a dummy local commit)

jugglinmike commented on tools/manifest/manifest.py:
```diff
@@ -400,8 +400,18 @@ def load(tests_root, manifest, types=None, meta_filters=None):
+
 def _load(logger, tests_root, manifest, types=None, meta_filters=None):
     # "manifest" is a path or file-like object.
+    manifest_path = manifest if isinstance(manifest, string_types) \
```
I share @jgraham's [mistrust of line continuations](https://github.com/web-platform-tests/wpt/pull/14004#discussion_r232222841), but [it doesn't look like we can disallow them with linting](https://pycodestyle.readthedocs.io/en/latest/intro.html#error-codes). Without that, I don't know if we can fairly reject the use of that syntax...

jgraham replied:
Yes we can :p Please don't use continuation lines, use additional parens instead.

jgraham commented on tools/manifest/manifest.py:
```diff
@@ -400,8 +400,18 @@ def load(tests_root, manifest, types=None, meta_filters=None):
+    if manifest_path in __load_cache:
+        return __load_cache[manifest_path]
+    # Cache at most one manifest.
+    __load_cache.clear()
```
Gecko uses two and servo uses three. I think we should allow one per url-base rather than one total.

Hexcles commented:
I addressed the review comments, but am having trouble debugging the `tools/wpt` failure on macOS -- I can't reproduce the error locally on my Macbook.

I'll try to dig deeper.

Hexcles committed:
Fix test_wpt.py

The test module uses the same temporary manifest in all test cases, but
some test cases modify the manifest, which causes inconsistency in the
cache. This commit changes the test setup to initialize a persistent
manifest on the module level, and each test case will then make a
temporary copy of it.

Also speed up tools/wpt tests on Azure by updating the manifest first.

Hexcles commented:
So my working theory is that `test_wpt.py` had a module-level fixture to create a temporary manifest, which means the manifest path was the same for all test cases, but some test cases need to modify the manifest, so caching by path failed. The workaround is to use a different temporary manifest for each test case (and the temporary manifest is copied from a "persistent" one, which is initialized in a module-level setup function to avoid having to generate the manifest every time). See my second commit. This seems to work. However, I don't really understand why the problem only demonstrated on macOS.

@jgraham PTAL again.

foolip commented:
FYI, I've rebased this branch to trigger "Travis CI - Pull Request", manually setting things right after the travis-ci.com transition in https://github.com/web-platform-tests/wpt/issues/14499 changed the name of the required check, which applies retroactively.
